### PR TITLE
feature targets compatibility

### PR DESCRIPTION
**Pull request by leandrodzp**
*When a target is created, all compatible targets are added to users.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -43,7 +43,23 @@ class User < ActiveRecord::Base
 
   has_many :targets, dependent: :destroy
 
+  has_many :user_compatible_targets
+  has_many :compatible_targets,
+           class_name: 'Target',
+           through: :user_compatible_targets,
+           source: 'target'
```
Comment (gonzaM) [inline app/models/user.rb]: Is this necessary? If yes, use `:target`

File: app/models/target.rb [hunk 2]
```diff
@@ -32,6 +38,26 @@ class Target < ApplicationRecord
...
+      [other_target.latitude, other_target.longitude]
+    )
+    ((distance < (radius + other_target.radius)) && (topic == other_target.topic))
+  end
+
+  def add_compatible_targets
+    not_mine = Target.where.not(user_id: user.id)
+    not_mine.each do |target|
+      if target.compatible_target?(self)
+        target.user.compatible_targets.concat(self)
```
Comment (gonzaM) [inline app/models/target.rb]: You can do: `user_compatible_targets.create!(user: target.user)`

File: app/models/target.rb [hunk 3]
```diff
@@ -32,6 +38,26 @@ class Target < ApplicationRecord
...
+        user.compatible_targets.concat(target)
```
Comment (gonzaM) [inline app/models/target.rb]: `target.user_compatible_targets.create!(user: user)`

File: app/models/target.rb [hunk 4]
```diff
@@ -32,6 +38,26 @@ class Target < ApplicationRecord
...
+      [latitude, longitude],
+      [other_target.latitude, other_target.longitude]
+    )
+    ((distance < (radius + other_target.radius)) && (topic == other_target.topic))
+  end
+
+  def add_compatible_targets
+    not_mine = Target.where.not(user_id: user.id)
+    not_mine.each do |target|
+      if target.compatible_target?(self)
```
Comment (gonzaM) [inline app/models/target.rb]: Use: `next unless target.compatible_target?(self)`

File: spec/requests/api/v1/targets/create_spec.rb [hunk 5]
```diff
@@ -26,6 +26,38 @@
           latitude:  params[:target][:latitude]
         )
       end
+      context 'when a compatible target exists' do
+        let(:user2)   { create(:user, confirmed_at: Time.zone.now) }
```
Comment (gonzaM) [inline spec/requests/api/v1/targets/create_spec.rb]: `user_2`

File: spec/requests/api/v1/targets/create_spec.rb [hunk 6]
```diff
@@ -26,6 +26,38 @@
...
+          )
+        end
+
+        it 'increments amount of compatible targets' do
+          expect do
+            post api_v1_targets_path, headers: auth_headers, params: params, as: :json
+          end.to change(user.compatible_targets, :count).by(1)
+        end
+
+        it 'increments amount of compatible targets of allready existing user' do
```
Comment (gonzaM) [inline spec/requests/api/v1/targets/create_spec.rb]: `'increments the amount of compatible targets for existing users'`

File: spec/requests/api/v1/targets/create_spec.rb [hunk 7]
```diff
@@ -26,6 +26,38 @@
...
+          expect do
+            post api_v1_targets_path, headers: auth_headers, params: params, as: :json
+          end.to change(user2.compatible_targets, :count).by(1)
+        end
+      end
+
+      context 'when a compatible target does not exist' do
```
Comment (gonzaM) [inline spec/requests/api/v1/targets/create_spec.rb]: Maybe a better message is: "when there are not compatible targets"

File: app/models/target.rb [hunk 8]
```diff
@@ -32,6 +38,25 @@ class Target < ApplicationRecord
...
+    distance = Geocoder::Calculations.distance_between(
+      [latitude, longitude],
+      [other_target.latitude, other_target.longitude]
+    )
+    ((distance < (radius + other_target.radius)) && (topic == other_target.topic))
+  end
+
+  def add_compatible_targets
+    not_mine = Target.where.not(user_id: user.id)
+    not_mine.each do |target|
```
Comment (mdornel) [inline app/models/target.rb]: Your solution works, but what if you have so many targets? You are going to iterate over the complete collection.

Maybe you can get a more accurate collection by querying right to the DB. Take a look to `earth` and `cube` postgres extension. [Here](https://salayhin.wordpress.com/2016/10/29/adding-a-radius-search-to-your-ruby-on-rails-5-application-and-postgresql/) is an example

File: app/models/target.rb [hunk 9]
```diff
@@ -22,6 +22,12 @@ class Target < ApplicationRecord
   belongs_to :user
   belongs_to :topic
 
+  has_many   :user_compatible_targets
+  has_many   :compatible_users,
+             class_name: 'User',
+             through: :user_compatible_targets,
+             source: 'user'
```
Comment (mdornel) [inline app/models/target.rb]: `source: :user` should work here

File: app/models/target.rb (hunk 8 above)
Comment (leandrodzp) [reply app/models/target.rb]: I know. The problem is that in this project there is no maximum radium so if I search the compatible targets in a radium maybe Im loosing some targets that are far away but are still compatible.

**Comment by leandrodzp** [comment]
Done :)

**Comment by leandrodzp** [comment]
Done :)

**Pull request merged**